src/x/os/interrupt.go

```diff
@@ -50,6 +69,33 @@ func (i *InterruptError) Error() string {
 	return i.interrupt
 }
 
+// WatchForInterrupt watches for interrupts in a non-blocking fashion and closes
+// the interrupted channel when an interrupt is found. Use this method to ensure
+// child goroutines get closed properly by passing them the interrupted channel.
+// If the interrupted channel is closed, then the goroutine knows to stop
+// its work.
+func WatchForInterrupt(logger *zap.Logger, opts InterruptOptions) func() {
+	interruptCh := opts.InterruptCh
+	closed := make(chan struct{})
+	go func() {
+		select {
+		case err := <-interruptCh:
+			logger.Warn("interrupt", zap.Error(err))
+			close(opts.InterruptedCh)
+		case <-closed:
+			logger.Info("interrupt watch stopped")
+			return
+		}
+	}()
+
+	var doOnce sync.Once
+	return func() {
+		doOnce.Do(func() {
```

Comment:
This sync.Once is just a bit of convenience so that startup methods can use it with `defer` so that it's invoked on early exits, but also can call it directly if they want to cancel the watch to switch to a synchronous wait for interrupt. Double closing causes panics, so ensure that we protect against that.